test/TimeLock.js

```diff
@@ -0,0 +1,180 @@
+'use strict';
+
+const chai = require('chai');
+const {
+  utils: { defaultAbiCoder, keccak256 },
+} = require('ethers');
+const { expect } = chai;
+const { ethers, network } = require('hardhat');
+const { HashZero } = ethers.constants;
+
+describe('TimeLock', async () => {
+  let owner;
+
+  let timeLockFactory;
+  let timeLock;
+
+  before(async () => {
+    [owner] = await ethers.getSigners();
+  });
+
+  beforeEach(async () => {
+    timeLockFactory = await ethers.getContractFactory('TimeLockTest', owner);
+
+    // 12 hour buffer
+    const buffer = 12 * 60 * 60;
+
+    timeLock = await timeLockFactory.deploy(buffer).then((d) => d.deployed());
+  });
+
+  it('should revert when scheduling a timelock with invalid hash', async () => {
+    const eta = 12 * 60 * 60;
+
+    await expect(
+      timeLock.scheduleSetNum(HashZero, eta),
+    ).to.be.revertedWithCustomError(timeLock, 'InvalidTimeLockHash');
+  });
+
+  it('should successfully schedule a timelock', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+  });
+
+  it('should set the timelock to the minimum time delay if the time delay provided is too low', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const minTimeDelay = 12 * 60 * 60;
+    const timeDelay = 2 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const minEta = block.timestamp + minTimeDelay;
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+
+    // allow for small block timestamp disparity
+    expect(Math.abs(etaTimeLock - minEta)).to.be.lessThanOrEqual(5);
```

Comment:
why is this needed? such workarounds usually mean there's an incorrect expectation by the test. Since the block timestamp when scheduling is `1` more than when the latest block was fetched, you just need to add 1 to minEta. Better yet, to make this network agnostic, get the block number from the tx receipt when scheduling the eta

Reply:
Changed the test to get the block timestamp from the tx receipt when scheduling the eta.